text/0000-semicolon-items.md

```diff
@@ -0,0 +1,448 @@
+
+### Retaining a uniform style
+
+To retain as uniform of a style possible in the Rust community,
+this RFC proposes that `rustfmt`, Rust's code formatting tool,
+should remove extraneous `;` since they are most likely left over
+while editing or as a frequently made mistake.
+No possibility of configuring this behaviour of `rustfmt` is proposed at this time.
+
+A warn by default lint will also be added to the compiler to further
+discourage extraneous `;`.
+
+## Leftover semicolons when refactoring
+
+Various common refactorings often leave behind extraneous semicolons, such as:
+
+* Replacing a closure with a `fn` item:
+  
+  ```rust
+  let add_two = |x| x + 2;
+  
+  // becomes:
+  fn add_two(x: u32) -> u32 {
+      x + 2
+  };
+  ```
+  
+* Adding a field to what was formerly a unit struct:
+
+  ```rust
+  struct UnitExpr;
+  
+  // becomes:
+  struct UnitExpr {
+      span: Span,
+  };
+  ```
+
+* Changing a tuple struct into a braced struct:
+
+  ```rust
+  struct Foo(String, usize);
+
+  // becomes:
+  struct Foo {
+      bar: String,
+      baz: usize,
+  };
+  ```
+ 
+The error emitted by the compiler in these circumstances is never
+indicative of a bug, but nonetheless disturbs the flow of writing
+and causes the user to have to wait and recompile.
+During that time, the user's train of thought can easily be lost.
+
+## Improving consistency with items in `fn` bodies
+
+[compiletest_rs]: https://github.com/laumann/compiletest-rs/blob/master/src/runtest.rs#L2585-L2660
+
+Incidentally, these extraneous semicolons are currently permitted on items
+defined inside of `fn` bodies, as a consequence of the fact that `;` in
+isolation is a valid *statement* (though not an item).
+The following, slightly modified example is from [compiletest_rs].
+
+```rust
+fn read2_abbreviated(mut child: Child) -> io::Result<Output> {
+    use std::mem::replace;
+    use read2::read2;
+
+    const HEAD_LEN: usize = 160 * 1024;
+    const TAIL_LEN: usize = 256 * 1024;
+
+    enum ProcOutput {
+        Full(Vec<u8>),
+        Abbreviated {
+            head: Vec<u8>,
+            skipped: usize,
+            tail: Box<[u8]>,
+        }
+    };
+
+    impl ProcOutput {
+        ...
+    };
+
+    ...
+}
+```
+
+By permitting semicolons as items outside of `fn` bodies,
+a modicum of consistency in Rust's grammar can be gained.
+
+## A frequent editing mistake
+
+For the authors of this RFC, who have written Rust code for many years,
+it still sometimes happens that they mistakenly place a semicolon after
+`struct Name {}`. As previously mentioned, such trivialities needlessly
+disrupt the editing process.
+
+## Habits from other languages
+
+Languages like C and C++ require that you write:
+
+```cpp
+struct Foo {
+    int x;
+    ...
+}; // <-- You can't omit the ;
+```
+
+Since you can't omit the semicolon from the definition above,
+that becomes habit. For people who use many languages, or are transitioning to
```

Comment:
Personal anecdote: I never accidentally write the extra `;` in Rust despite years of C++ background, I rather forget to write it in C++ after Rust.

Reply:
That's the primary source of my belief that the Rust way is superior and we should keep it as is rather than pulling syntactic garbage into the language.

Reply:
That is fair, but I believe there has been enough anecdote to the contrary in this thread. Some people have to switch between languages and can't solely write in Rust, as unfortunate as that is.

I think "garbage" is harsh; this is no different than supporting `;` inside `fn` bodies imo.

Reply:
>Some people have to switch between languages and can't solely write in Rust, as unfortunate as that is.

FWIW, I'm one of those people, I write C++ at work.

Reply:
With the lint provided, you can always `#![forbid(redundant_semicolon)]` and you'll actually be more strict about your code in Rust than before this RFC :)